### DISCUSSION TITLE
OrtApi Globals in C#

### DISCUSSION THREAD

**Discussion by thoron** [microsoft/onnxruntime#13307] [API Q&A] [answered]
Hi everyone,

I am missing some of the C/java bindings in C#. Specifically SpinLock control and global threading options.
They already exist in the `OrtApi` struct in the wrapper: https://github.com/microsoft/onnxruntime/blob/a83a9ed6b01374a5e1e07b46dbf2fb8e3c407483/csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.shared.cs#L183

Could it be considered adding them to the `OrtEnv` class?
Similar to the java implementation: https://github.com/microsoft/onnxruntime/blob/0757d5133428a9c0afb65bc0f74b3ced81d77cd4/java/src/main/java/ai/onnxruntime/OrtEnvironment.java#L430

I suggest exposing the threading options same as for java:
```java
    OrtEnvironment.ThreadingOptions threadOpts = new OrtEnvironment.ThreadingOptions();
    threadOpts.setGlobalInterOpNumThreads(2);
    threadOpts.setGlobalIntraOpNumThreads(2);
    threadOpts.setGlobalDenormalAsZero();
    threadOpts.setGlobalSpinControl(true);
```
(from https://github.com/microsoft/onnxruntime/blob/main/java/src/test/java/ai/onnxruntime/EnvironmentThreadPoolTest.java)

Or expanding the `OrtEnv` with additional methods:
```csharp
        public void SetGlobalSpinControl(bool allowSpin)
        {
            NativeApiStatus.VerifySuccess(NativeMethods.OrtSetGlobalSpinControl(Handle, allowSpin));
        }
```

Using it as:
```csharp
OrtEnv.Instance().SetGlobalInterOpNumThreads(2);
OrtEnv.Instance().SetGlobalIntraOpNumThreads(2);
OrtEnv.Instance().SetGlobalSpinControl(true);
```

Best regards,
Thomas

**Comment by yuslepukhin** [comment]
Cc: @RandySheriffH

**Comment by thoron** [comment, marked as answer]
Discussion moved to issue #13743 and fixed in PR #13822
